src/sim.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python3
+import math
+import numpy as np
+import pandas as pd
+import salabim as sim
+import sys
+
+
+#  based on SimPy example model
+class Bus(sim.Component):
+    """
+    TODO
+    """
+    def __init__(self, start_time, end_time, trip_power, charging_rate, *args, **kwargs):
+        sim.Component.__init__(self, *args, **kwargs)
+        self.start_time    = start_time
+        self.end_time      = end_time
+        self.trip_power    = trip_power
+        self.charging_rate = charging_rate
+
+    def process(self):
+        """TODO
+        """
+        #Trip starts when this component is activated
+        yield self.request(bus_depot)
+        yield self.hold(till=self.end_time)
+        yield self.request((energy, self.trip_power))
+        yield self.hold(duration=self.trip_power/self.charging_rate*3600)
+        self.release(bus_depot)
+
+
+
+class Dispatcher:
```

Comment:
Let me see if I understand this. Dispatcher takes in the df of trips, and all the constraints set across a homogeneous fleet.  It then finds all the trips that need to performed by a discrete "bus" (identified by the `block_id` and groups them. We then schedule all the trips for that block create a bus for it. I dont quite grok how the logic for the final `for` loop in `schedule_block` works. Lets talk through that.

Reply:
Mostly right. The dispatcher identifies `block_id` and then breaks it apart into segments that are serviceable by buses of the defined capacity.

```
trips['trip_seg'] = (trips.cumdist/1000/maxdist).astype(np.int)
```
Creates an integer-valued entry which indicates which parts of a trip can be serviced with a common bus.